### PR TITLE
Match native aot trim annotations between reference and implementation assemblies

### PR DESCRIPTION
**Pull request by LakshanF**
Remove `RequiresDynamicCodeAttribute `from the API compatibility exclusion list to assess the impact on core runtime libraries.

### CODE REVIEW

File: src/mono/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs [hunk 1]
```diff
@@ -193,48 +195,56 @@ public static int GetExceptionCode()
             throw new PlatformNotSupportedException();
         }
 
+        [RequiresDynamicCode("Marshalling code for the object might not be available")]
```
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs:198]: Hmm, not sure about these. Does this add any value?

Comment (LakshanF) [reply src/mono/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs:198]: The `APICompat` check fails if a public implementation API doesn't match the reference API (and vice versa). 

The `CoreRT `implementation required this API to be annotated, that in turn required the reference API to be annotated which is shared by all implementations, which required other runtime implementations to be annotated.

Comment (MichalStrehovsky) [reply src/mono/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs:198]: Yeah, these APIs are already marked obsolete and there will be _some_ warning for them. But the AOT warnings generally follow the philosophy of trimming warnings: if there's a possible behavioral difference, there should be a warning.

These APIs allow running marshalling on anything, then read a byte out the marshalled representation, and destroy it. They can't work if we can't build the marshaller dynamically at runtime (and it wasn't precompiled).

**Review by marek-safar — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
Cc @eerhardt

**Review by MichalStrehovsky — APPROVED** [review]
Yep, the failures look unrelated. The NativeAOT failure is surprising, I'll need to check on that.

File: src/libraries/System.Reflection.Emit.Lightweight/ref/System.Reflection.Emit.Lightweight.cs [hunk 2]
```diff
@@ -30,13 +30,21 @@ public sealed partial class DynamicILInfo
     }
     public sealed partial class DynamicMethod : System.Reflection.MethodInfo
     {
+        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
```
Comment (eerhardt) [inline src/libraries/System.Reflection.Emit.Lightweight/ref/System.Reflection.Emit.Lightweight.cs:33]: Do we want to use `"native AOT"` in the public message shown to users? That seems a bit too specific about the technology here. That's why we didn't name the attribute "NativeAOTUnsupported".

```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```

**Review by eerhardt — APPROVED** [review]
Just one minor comment about the message. LGTM otherwise.

File: src/libraries/System.Reflection.Emit.Lightweight/ref/System.Reflection.Emit.Lightweight.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Reflection.Emit.Lightweight/ref/System.Reflection.Emit.Lightweight.cs]: Note that this comment applies throughout the entire change. Not just this one place.

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs [hunk 3]
```diff
@@ -19,9 +19,9 @@ public sealed partial class AssemblyBuilder : System.Reflection.Assembly
         public override string Location { get { throw null; } }
         public override System.Reflection.Module ManifestModule { get { throw null; } }
         public override bool ReflectionOnly { get { throw null; } }
-        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:22]: These two aren't for creating a DynamicMethod:

```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs [hunk 4]
```diff
@@ -19,9 +19,9 @@ public sealed partial class AssemblyBuilder : System.Reflection.Assembly
...
         public static System.Reflection.Emit.AssemblyBuilder DefineDynamicAssembly(System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access) { throw null; }
-        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:24]: These two aren't for creating a DynamicMethod:

```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs [hunk 5]
```diff
@@ -139,7 +139,7 @@ private void InitManifestModule()
 
         #region DefineDynamicAssembly
 
-        [RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs:142]: These two aren't for creating a DynamicMethod:

```suggestion
        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs [hunk 6]
```diff
@@ -150,7 +150,7 @@ public static AssemblyBuilder DefineDynamicAssembly(AssemblyName name, AssemblyB
                                                  null);
         }
 
-        [RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs:153]: These two aren't for creating a DynamicMethod:

```suggestion
        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/mono/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 7]
```diff
@@ -236,7 +236,7 @@ public override bool ReflectionOnly
             get { return base.ReflectionOnly; }
         }
 
-        [RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:239]: These two aren't for creating a DynamicMethod:

```suggestion
        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/mono/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 8]
```diff
@@ -245,7 +245,7 @@ public static AssemblyBuilder DefineDynamicAssembly(AssemblyName name, AssemblyB
             return new AssemblyBuilder(name, access);
         }
 
-        [RequiresDynamicCode("Generating new code at runtime is not supported with native AOT.")]
+        [RequiresDynamicCode("Creating a DynamicMethod requires dynamic code.")]
```
Comment (eerhardt) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:248]: These two aren't for creating a DynamicMethod:

```suggestion
        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs (hunk 5 above)
Comment (LakshanF) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs]: Thanks

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**